roomba/password.py

```diff
@@ -197,35 +199,101 @@ def get_password_from_roomba(self, addr):
-        
-            while len(data) < 37:
-                data_received = wrappedSocket.recv(1024)
-                data+= data_received
-                if len(data_received) == 0:
-                    self.log.info("socket closed")
-                    break
-                
-            wrappedSocket.close()
-            return data
-            
-        except socket.timeout as e:
-            self.log.error('Connection Timeout Error (for {}): {}'.format(addr, e))
-        except (ConnectionRefusedError, OSError) as e:
-            if e.errno == 111:      #errno.ECONNREFUSED
-                self.log.error('Unable to Connect to roomba at ip {}, make sure nothing else is connected (app?), '
-                               'as only one connection at a time is allowed'.format(addr))
-            elif e.errno == 113:    #errno.No Route to Host
-                self.log.error('Unable to contact roomba on ip {} is the ip correct?'.format(addr))
+            try:
+                wrappedSocket.send(packet)
+                self.log.debug('Waiting for data')
+
+                while len(data) < 37:
+                    data_received = wrappedSocket.recv(1024)
+                    data+= data_received
+                    if len(data_received) == 0:
+                        self.log.info("socket closed")
+                        break
+            finally:
+                wrappedSocket.close()
+        except Exception as e:
+            data = b''
+            if isinstance(e, socket.timeout):
+                self.log.error('Connection Timeout Error (for {}): {}'.format(addr, e))
+            elif isinstance(e, ConnectionRefusedError) or isinstance(e, OSError):
+                if e.errno == 111:      #errno.ECONNREFUSED
+                    self.log.error('Unable to Connect to roomba at ip {}, make sure nothing else is connected (app?), '
+                                   'as only one connection at a time is allowed'.format(addr))
+                elif e.errno == 113:    #errno.No Route to Host
+                    self.log.error('Unable to contact roomba on ip {} is the ip correct?'.format(addr))
+                else:
+                    self.log.error("Connection Error (for {}): {}".format(addr, e))
             else:
-                self.log.error("Connection Error (for {}): {}".format(addr, e))
+                self.log.exception(e)
+
+            self.log.error('Unable to get password from roomba')
+            return None
+
+        if len(data) <= 7:
+            return None
+
+        # Convert password to str
+        password = data[7:]
+        if b'\x00' in password:
+            password = password[:password.find(b'\x00')]  # for i7 - has null termination
+        return str(password.decode())
+
+    def generate_new_password(self):
+        alphabet = string.ascii_letters + string.digits
+        password = ":1:{}:".format(int(time.time()))
+        while len(password) < 30:
+            password += random.choice(alphabet)
```

Comment:
I am not sure how much is the password relevant to security, but I'd avoid using the [random module](https://docs.python.org/3/library/random.html), as they warn it is not suitable for those purposes:

> The pseudo-random generators of this module should not be used for security purposes. For security or cryptographic uses, see the [secrets](https://docs.python.org/3/library/secrets.html#module-secrets) module.

It seems that [secrets.choice](https://docs.python.org/3/library/secrets.html#module-secrets) is a suitable replacement for random.choice.

The secrets module has also some other methods for generating tokens (so you don't have to add the characters one-by-one), but it seems it is not that easy to generate alphanumeric string.